package/MDAnalysis/analysis/density.py

```diff
@@ -180,11 +177,259 @@
+                 metadata=None, padding=2.0,
+                 parameters=None, gridcenter=None,
+                 xdim=None, ydim=None, zdim=None):
+        u = atomgroup.universe
+        super(DensityAnalysis, self).__init__(u.trajectory)
+        self._atomgroup = atomgroup
+        self._delta = delta
+        self._metadata = metadata
+        self._padding = padding
+        self._parameters = parameters
+        self._gridcenter = gridcenter
+        self._xdim = xdim
+        self._ydim = ydim
+        self._zdim = zdim
+
+    def _prepare(self):
+        coord = self._atomgroup.positions
+        if self._gridcenter is not None:
+            # Issue 2372: padding is ignored, defaults to 2.0 therefore warn
+            if self._padding > 0:
+                msg = ("Box padding (currently set at {0}) "
+                       "is not used in user defined grids.".format(self._padding))
+                warnings.warn(msg)
+                logger.warning(msg)
+            # Generate a copy of smin/smax from coords to later check if the
+            # defined box might be too small for the selection
+            smin = np.min(coord, axis=0)
+            smax = np.max(coord, axis=0)
+            # Overwrite smin/smax with user defined values
+            smin, smax = _set_user_grid(self._gridcenter, self._xdim,
+                                        self._ydim, self._zdim, smin, smax)
+        else:
+            # Make the box bigger to avoid as much as possible 'outlier'. This
+            # is important if the sites are defined at a high density: in this
+            # case the bulk regions don't have to be close to 1 * n0 but can
+            # be less. It's much more difficult to deal with outliers.  The
+            # ideal solution would use images: implement 'looking across the
+            # periodic boundaries' but that gets complicated when the box
+            # rotates due to RMS fitting.
+            smin = np.min(coord, axis=0) - self._padding
+            smax = np.max(coord, axis=0) + self._padding
+        BINS = fixedwidth_bins(self._delta, smin, smax)
+        arange = np.transpose(np.vstack((BINS['min'], BINS['max'])))
+        bins = BINS['Nbins']
+        # create empty grid with the right dimensions (and get the edges)
+        grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins,
+                                     range=arange, normed=False)
+        grid *= 0.0
+        self._grid = grid
+        self._edges = edges
+        self._arange = arange
+        self._bins = bins
+        #: density as a :class:`Density`
+        self.density = None
+
+    def _single_frame(self):
+        h, _ = np.histogramdd(self._atomgroup.positions,
+                              bins=self._bins, range=self._arange,
+                              normed=False)
+        # manually reduce (not yet part of AnalysisBase but exists in pmda.parallel.AnalysisBase)
+        self._reduce(self._grid, h)
```

Comment:
Probably being silly here so please ignore if that's the case. Is there a plan to introduce _reduce to `AnalysisBase` in the future? Otherwise, it's not clear to me how the extra `if/else` isn't introducing an unnecessary couple of extra CPU cycles per iteration.

Reply:
I think for 2.0 I'd like to rewrite AnalysisBase to match pmda so that we don't maintain two subtly different implementations. When we started with the serial version, we didn't really understand what was necessary to make the parallel one work. I opened #2542 .

Reply:
I commented out the _reduce code and replaced it with the simple serial version.